src/codegen/compile.ml

```diff
@@ -8006,14 +8006,33 @@ and compile_exp (env : E.t) ae exp =
       SR.Vanilla,
       ContinuationTable.size env ^^ Prim.prim_word32toNat env
 
+    | OtherPrim "rts_stable_var_footprint_query", [] ->
+      SR.Vanilla,
+      Var.get_val_vanilla env ae "$__stable_variable_footprint"
+
+    | OtherPrim "☠rts_stable_vars_size☠", [e] ->
+      SR.UnboxedWord64,
+      let ts = [e.note.Ir_def.Note.typ] in
+      let tydesc = Serialization.type_desc env ts in
+      let tydesc_len = Int32.of_int (String.length tydesc) in
+      compile_exp_vanilla env ae e ^^
+      Serialization.buffer_size env (Type.seq ts) ^^
+      G.i Drop ^^
+      compile_add_const tydesc_len  ^^
+      G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32))
+
+    | OtherPrim "⏰rts_perform_gc⏰", [] ->
```

Comment:
I don't think this should be a primitive, otherwise it could easily be called with a non-empty stack and collect data that is actually live by only being reachable from the stack.

I wonder if it would be safter to add the call to collect not to the body the preupgrade Motoko system method, but to the wasm only canister_preupgrade (that calls the former).

But perhaps I'm being anal about this and just not exposing it in `Prim` is enough if we are careful.

Reply:
I think it protection would be worth it. I'm sure someone will find this and try to invoke it to force gcs at the wrong time.
Alternatively, to avoid the prim, refactor as I suggest below.